src/dfx/src/lib/identity/identity_manager.rs

```diff
@@ -33,6 +30,38 @@ fn default_identity() -> String {
 #[derive(Clone, Debug, Default, Serialize, Deserialize)]
 pub struct IdentityConfiguration {
     pub hsm: Option<HardwareIdentityConfiguration>,
+
+    /// If the identity's .pem file is encrypted this contains everything (except the password) to decrypt the file.
+    pub encryption: Option<EncryptionConfiguration>,
+}
+
+/// The information necessary to de- and encrypt (except the password) the identity's .pem file
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct EncryptionConfiguration {
+    /// Salt used for deriving the key from the password
+    pub pw_salt: String,
+
+    /// 96 bit Nonce used to decrypt the file
+    pub file_nonce: Vec<u8>,
+}
+
+impl EncryptionConfiguration {
+    /// Generates a random salt and nonce. Use this for every new identity
+    pub fn new() -> DfxResult<Self> {
+        let mut nonce: [u8; 12] = [0; 12];
+        let mut salt: [u8; 32] = [0; 32];
+        let sr = rand::SystemRandom::new();
+        sr.fill(&mut nonce)?;
```

Comment:
As I understand it, this `EncryptionConfiguration` is going to be used to encrypt multiple different identities, at least potentially. If that's the case this is bad - if a nonce is reused with GCM, confidentiality is lost. A nonce/key pair must only be used a single time.

I'm not totally familiar with how this system works but some things to consider

* It's possible to generate both the AES key and the GCM nonce using Argon2
* Argon2 supports auxiliary inputs such as "associated data" (`ad` in the API). If you use Argon2 with the same password but different associated datas, you will generate new effectively unique keys for each `ad`. If you set the `ad` to the serialization of the identity name then you will get a unique key/nonce per file.

This *still* has a problem if you re-encrypt a *different* data under the same key/nonce. AIUI this doesn't happen here - the identity key is generated, and then later decrypted many times, but is never modified/rewritten.

Reply:
Thanks for the comments! `EncryptionConfiguration` is separate per identity. See example:
<img width="583" alt="image" src="https://user-images.githubusercontent.com/14757975/158543556-0a6c9433-d5e4-443b-8caf-2f4ed395c481.png">

Using the identity name to derive anything (or using it as associated data) won't work since identities can be renamed at will (at least at the moment).

Using Argon2 to generate key and nonce is certainly possible. But what would you use as the input? Using `SystemRandom` is certainly possible, but I don't quite see how that would be a randomness upgrade (applying a deterministic function to an input shouldn't change it's entropy, right?). Using the password to derive them will lead to key/nonce reuse .